### PR TITLE
Add support for both custom quote and escape characters

### PR DESCRIPTION
**Pull request by galenandrew**
**Added support for both _custom quote character_ and _custom escape character_.**
_related to #28_

This Pull Request consists of deeper changes to the parsing logic (specifically around parsing cell values), but provides greater flexibility when parsing non-standard CSV files.

For example, a CSV formatted as such…
```
a,b,c
1,"some \"escaped\" value",2
3,4,5
```
Can now be parsed to…
```javascript
[
  {a: '1', b: 'some "escaped" value', c: '2'},
  {a: '3', b: '4', c: '5'}
]
```

_NOTE: I tried to test a handful of common scenarios to ensure backwards compatibility and feel good with the results, but please review and make sure this functions as desired._

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -56,8 +57,13 @@ Parser.prototype._transform = function (data, enc, cb) {
   }
 
   for (var i = start; i < buf.length; i++) {
-    if (buf[i] === this.escape) this._escaped = !this._escaped
-    if (!this._escaped) {
+    // non-escaped quote (quoting the cell)
+    if (buf[i] === this.quote && buf[i - 1] !== this.escape) {
```
Comment (mafintosh) [inline index.js]: could you check that `i > 0` after the `&&`? otherwise we're gonna get a deopt from v8

File: index.js [hunk 2]
```diff
@@ -102,19 +108,19 @@ Parser.prototype._online = function (buf, start, end) {
...
   var offset = start
 
   for (var i = start; i < end; i++) {
-    if (buf[i] === this.escape) { // "
-      if (i < end - 1 && buf[i + 1] === this.escape) { // ""
-        i++
-      } else {
-        isEscaped = !isEscaped
-      }
+    if (buf[i] === this.escape && buf[i + 1] === this.quote) {
```
Comment (mafintosh) [inline index.js]: same as above but `i + 1 < end`

File: index.js [hunk 3]
```diff
@@ -102,19 +108,19 @@ Parser.prototype._online = function (buf, start, end) {
...
+      i++
+      continue
+    } else if (buf[i] === this.quote && buf[i - 1] !== this.escape) {
```
Comment (mafintosh) [inline index.js]: same as above but `i - 1 >= start`

File: index.js [hunk 4]
```diff
@@ -166,13 +172,15 @@ Parser.prototype._emit = function (Row, cells) {
...
+  // remove quotes from quoted cells
+  if (buf[start] === this.quote && buf[end - 1] === this.quote) {
     start++
     end--
   }
 
   for (var i = start, y = start; i < end; i++) {
-    if (buf[i] === this.escape && buf[i + 1] === this.escape) i++
+    // check for escape characters and skip them
+    if (buf[i] === this.escape && buf[i + 1] === this.quote) i++
```
Comment (mafintosh) [inline index.js]: i guess there should be a bound check here as well. not sure why we didn't have that before

**Comment by mafintosh** [comment]
Looks good except for my comments. The parser is pretty performance critical so I wanna avoid deopts even though it makes the code a bit more c like :/

**Comment by galenandrew** [comment]
@mafintosh thanks for the feedback! After adding the additional bound checks, a few of the tests broke so I'll dig deeper and resolve accordingly.

**Comment by galenandrew** [comment]
@mafintosh apologizes for the delay getting around to this. Please see my updated diffs with bound checks in place.

After adding the initial bound checks, the tests for empty cells were failing so I refactored the checks for starting and ending quotes for cells.

Let me know if you see anything else that needs updating. Thanks!

**Comment by galenandrew** [comment]
@mafintosh any update on this?

**Comment by mafintosh** [comment]
@galenandrew sorry missed your update. this looks good. could you remove the version bump from pkg json? it's easier if i do that so i can tag the release etc.

**Comment by mafintosh** [comment]
@galenandrew hmm.. i see that those are my own commits. wonder why they show up as additions

**Comment by galenandrew** [comment]
I think I did a downstream merge which is why they show up in the history of this branch.

**Comment by mafintosh** [comment]
I cherry-picked the commits over in master to not mess up the history. Released as 1.9.0. Thanks!

**Comment by galenandrew** [comment]
Thanks!!
